### PR TITLE
misc: Convert miscellenous OPA usage to MPL atomics

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

Converts the rest of the OpenPA usage to MPL_atomics.

This is a split from #4377. It gathers most of the clean atomic conversions.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/warnings/quick
test:jenkins/ch3/most
test:jenkins/ch4/most

test:jenkins/custom --with-ch4-posix-eager-modules=fbox
netmod: ch4:ofi

File: src/mpid/ch4/shm/posix/eager/fbox/fbox_types.h [hunk 1]
```diff
@@ -14,20 +14,18 @@
 #include <mpidimpl.h>
 #include "mpidu_init_shm.h"
 
-#define MPIDI_POSIX_FBOX_DATA_LEN  (16 * 1024 - sizeof(uint64_t) - 2 * sizeof(int))
-#define MPIDI_POSIX_FBOX_THRESHOLD (MPIDI_POSIX_FBOX_DATA_LEN)
+#define MPIDI_POSIX_FBOX_SIZE       16 * 1024
```
Comment (pavanbalaji) [inline src/mpid/ch4/shm/posix/eager/fbox/fbox_types.h]: Can you put the number in parenthesis?

File: src/include/mpir_refcount_pobj.h [hunk 2]
```diff
@@ -10,45 +10,24 @@
...
 {
     int complete;
 
-    complete = (0 == OPA_load_int(cc_ptr));
+    complete = (0 == MPL_atomic_load_int(cc_ptr));
     if (complete) {
-        MPL_VG_ANNOTATE_HAPPENS_AFTER(cc_ptr);
-        OPA_read_barrier();
+        /* REVIEW: is this necessary? */
+        MPL_atomic_read_barrier();
```
Comment (pavanbalaji) [inline src/include/mpir_refcount_pobj.h]: This read barrier should not be necessary.  The atomic load_int above should already provide both a load and a store barrier (https://en.cppreference.com/w/cpp/atomic/memory_order).

File: src/mpid/ch4/shm/posix/eager/fbox/fbox_types.h (hunk 1 above)
Comment (hzhou) [reply src/mpid/ch4/shm/posix/eager/fbox/fbox_types.h]: done.

File: src/include/mpir_refcount_pobj.h (hunk 2 above)
Comment (hzhou) [reply src/include/mpir_refcount_pobj.h]: removed.

**Comment by hzhou** [comment]
test:mpich/warnings

**Comment by hzhou** [comment]
test:jenkins/ch4/most

**Comment by hzhou** [comment]
Something went wrong with Jenkins all of a sudden.

test:jenkins/warnings/quick

**Comment by hzhou** [comment]
test:whitespace

**Comment by hzhou** [comment]
test:jenkins/ch4/most

**Comment by hzhou** [comment]
test:whitespace

**Comment by hzhou** [comment]
test:whitespace

**Comment by hzhou** [comment]
@pavanbalaji Ready for your review again.

File: src/mpid/ch4/shm/posix/eager/fbox/fbox_types.h [hunk 3]
```diff
@@ -14,20 +14,18 @@
...
+    MPL_atomic_int_t data_ready;
     int is_header;
     size_t payload_sz;
-
-    uint8_t payload[MPIDI_POSIX_FBOX_DATA_LEN];
-
+    uint8_t payload[];
 } MPIDI_POSIX_fastbox_t;
 
+#define MPIDI_POSIX_FBOX_DATA_LEN  (MPIDI_POSIX_FBOX_SIZE - sizeof(MPIDI_POSIX_fastbox_t))
```
Comment (pavanbalaji) [inline src/mpid/ch4/shm/posix/eager/fbox/fbox_types.h]: You cannot do `sizeof` on structs that have flexible arrays.

Comment (hzhou) [reply src/mpid/ch4/shm/posix/eager/fbox/fbox_types.h]: Yes, you can. I believe that is the signature feature. According to wikipedia https://en.wikipedia.org/wiki/Flexible_array_member:

> The sizeof operator on such a struct gives the size of the structure as if the flexible array member was empty.

Comment (pavanbalaji) [reply src/mpid/ch4/shm/posix/eager/fbox/fbox_types.h]: Interesting.  I didn't know that.

**Pull request merged**